community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledConversionUtils.java

```diff
@@ -44,76 +43,72 @@
+import org.neo4j.values.virtual.ListValue;
+import org.neo4j.values.virtual.MapValue;
+import org.neo4j.values.virtual.NodeValue;
+import org.neo4j.values.virtual.RelationshipValue;
+import org.neo4j.values.virtual.VirtualNodeValue;
+import org.neo4j.values.virtual.VirtualRelationshipValue;
+import org.neo4j.values.virtual.VirtualValues;
 
 import static java.lang.String.format;
+import static org.neo4j.values.SequenceValue.IterationPreference.RANDOM_ACCESS;
+import static org.neo4j.values.storable.Values.NO_VALUE;
 
 // Class with static methods used by compiled execution plans
 @SuppressWarnings( "unused" )
 public abstract class CompiledConversionUtils
 {
     public static boolean coerceToPredicate( Object value )
     {
-        if ( value == null )
+        if ( value == null || value == Values.NO_VALUE )
         {
             return false;
         }
+        if ( value instanceof BooleanValue )
+        {
+            return ((BooleanValue) value).booleanValue();
+        }
         if ( value instanceof Boolean )
         {
             return (boolean) value;
         }
+        if ( value instanceof ArrayValue )
+        {
+            return ((ArrayValue) value).length() > 0;
+        }
         if ( value.getClass().isArray() )
         {
             return Array.getLength( value ) > 0;
         }
         throw new CypherTypeException( "Don't know how to treat that as a predicate: " + value.toString(), null );
     }
 
-    public static Collection<?> toCollection( Object value )
+    public static Set<?> toSet( Object value )
     {
-        if ( value == null )
-        {
-            return Collections.emptyList();
-        }
-        else if ( value instanceof Collection<?> )
-        {
-            return (Collection<?>) value;
-        }
-        else if ( value instanceof LongStream )
-        {
-            LongStream stream = (LongStream) value;
-            return stream.boxed().collect( Collectors.toList() );
-        }
-        else if ( value instanceof IntStream )
+        if ( value == null || value instanceof NoValue )
         {
-            IntStream stream = (IntStream) value;
-            return stream.boxed().collect( Collectors.toList() );
-        }
-        else if ( value instanceof DoubleStream )
-        {
-            DoubleStream stream = (DoubleStream) value;
-            return stream.boxed().collect( Collectors.toList() );
+            return Collections.emptySet();
         }
-        else if ( value.getClass().isArray() )
+        else if ( value instanceof SequenceValue )
         {
-            int len = Array.getLength( value );
-            ArrayList<Object> collection = new ArrayList<>( len );
-            for ( int i = 0; i < len; i++ )
+            SequenceValue sequenceValue = (SequenceValue) value;
+            Iterator<AnyValue> iterator = sequenceValue.iterator();
+            Set<AnyValue> set = sequenceValue.iterationPreference() == RANDOM_ACCESS ?
```

Comment:
If `RANDOM_ACCESS` we can save on creating the iterator as well, so something like

```
if ( sequenceValue.iterationPreference() == RANDOM_ACCESS )
{
  for ( int i = 0; i < sequenceValue.length(); i++ )
  {
    set.add( sequenceValue.value( i ) );
  }
}
else
{
  //same as above
}
```